### ISSUE TITLE
Make prop-types work when renaming this.props

### ISSUE THREAD

**Issue by JelteF** [yannickcr/eslint-plugin-react#810]
I think it is handy for non stateless components with lots of properties to do something like this:

```jsx
class TestComponent extends Component {
    render() {
        let props = this.props;
        return (
            <div>Hello {props.name} </div>
        );
    }
}
```

This does not trigger the [prop-types](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/prop-types.md) warning though. Would it be possible to make that work?

**Comment by yannickcr** [comment]
Since we only use the AST it's very difficult to follow variable reassignments and making it would add a lot of complexity to the rules.

In your case I suggest you to use destructuring instead which is handled by the rule:

```jsx
class TestComponent extends Component {
    render() {
        let {name} = this.props;
        return (
            <div>Hello {name} </div>
        );
    }
}
```

**Comment by JelteF** [comment]
Why would this be difficult? The only thing that is required is that a list of aliases should be kept.

**Comment by mydearxym** [comment]
@yannickcr   this is not work ,  if the name is a object , will still get warning

**Comment by yannickcr** [comment]
@mydearxym What warning are you taking about? With destructuring you correctly get the `'name' is missing in props validation` warning, which does not trigger with the first pattern.

@JelteF if you have a solution feel free to submit a PR 😃

**Comment by mydearxym** [comment]
@yannickcr  what i mean is if name if a object : 

```js
class TestComponent extends Component {
    render() {
        let {name: { firstName }} = this.props;
        return (
            <div>Hello {firstName} </div>
        );
    }
}

TestComponent.propTypes = {
  name: React.PropTypes.shape({
    firstName: React.PropTypes.string   // warning , defined but never used 
  })
};

```

**Comment by yannickcr** [comment]
@mydearxym ok, seems you are talking about the `no-unused-prop-types` rule, this issue is about the `prop-types` one 😉 

(I am aware of the false positives in `no-unused-prop-types`, I hope to work on it this week)

**Comment by mydearxym** [comment]
yes is about `no-unused-prop-types`, my mistake 😉
